ldap/servers/slapd/ldapi.c

```diff
@@ -0,0 +1,409 @@
+            if (LDAP_SUCCESS == result)
+                slapi_pblock_get(
+                    search_pb,
+                    SLAPI_PLUGIN_INTOP_SEARCH_ENTRIES,
+                    &entries);
+
+            if (entries) {
+                /* zero or multiple entries fail */
+                if (entries[0] && 0 == entries[1]) {
+                    /* observe account locking */
+                    ret = slapi_check_account_lock(
+                        0, /* pb not req */
+                        entries[0],
+                        0, /* no response control */
+                        0, /* don't check password policy */
+                        0  /* don't send ldap result */
+                        );
+
+                    if (0 == ret) {
+                        auth_dn = slapi_ch_strdup(slapi_entry_get_ndn(entries[0]));
+                        bind_credentials_set_nolock(
+                            conn,
+                            SLAPD_AUTH_OS,
+                            auth_dn,
+                            NULL, NULL,
+                            NULL, entries[0]);
+                    }
+                }
+            }
+
+        entry_map_free:
+            /* auth_dn consumed by bind creds set */
+            slapi_free_search_results_internal(search_pb);
+            slapi_pblock_destroy(search_pb);
+            slapi_ch_free_string(&filter);
+            slapi_ch_free_string(&utype);
+            slapi_ch_free_string(&gtype);
+            slapi_ch_free_string(&base_dn);
+        }
+
+        /*
+         * We map the current process uid also to directory manager.
+         * This is secure as it requires local machine OR same-container volume
+         * access and the correct uid access. If you have access to the uid/gid
+         * and are on the same machine you could always just reset the rootdn hashes
+         * anyway ... so this is no reduction in security.
+         */
+
+        if (ret && (0 == uid || proc_uid == uid || proc_gid == gid)) {
+            /* map unix root (uidNumber:0)? */
+            char *root_dn = config_get_ldapi_root_dn();
+
+            if (root_dn) {
+                Slapi_PBlock *entry_pb = NULL;
+                Slapi_DN *edn = slapi_sdn_new_dn_byref(slapi_dn_normalize(root_dn));
+                Slapi_Entry *e = NULL;
+
+                /* root might be locked too! :) */
+                ret = slapi_search_get_entry(&entry_pb, edn, 0, &e, (void *)plugin_get_default_component_id());
+                if (0 == ret && e) {
+                    ret = slapi_check_account_lock(
+                        0, /* pb not req */
+                        e,
+                        0, /* no response control */
+                        0, /* don't check password policy */
+                        0  /* don't send ldap result */
+                        );
+
+                    if (1 == ret)
+                        /* sorry root,
+                         * just not cool enough
```

Comment:
Consider adding "😎" to this comment for impact.

Reply:
I did not add that comment (wish I had) :-)

Reply:
Nothing stopping you from improving the comment for future generations :)

Reply:
Fair enough